test/integration/master/kms_transformation_test.go

```diff
@@ -140,22 +122,67 @@ func TestKMSProvider(t *testing.T) {
 	}
 
 	// Secrets should be un-enveloped on direct reads from Kube API Server.
-	s, err := test.restClient.CoreV1().Secrets(testNamespace).Get(testSecret, metav1.GetOptions{})
+	mustReadSecret(t, test, true)
+	test.printMetrics()
+}
+
+func mustSetupTest(t *testing.T) (*base64Plugin, *transformTest) {
+	pluginMock, err := NewBase64Plugin()
+	if err != nil {
+		t.Fatalf("failed to create mock of KMS Plugin: %v", err)
+	}
+
+	test, err := newTransformTest(t, kmsConfigYAML)
+	if err != nil {
+		pluginMock.cleanUp()
+		t.Fatalf("failed to start KUBE API Server with encryptionConfig\n %s", kmsConfigYAML)
+	}
+
+	// As part of newTransformTest a new secret was created, so KMS Mock should have been exercised by this point.
+	if len(pluginMock.errorChan) != 0 {
+		pluginMock.cleanUp()
+		test.cleanUp()
+		t.Fatalf("KMSPlugin failed while serving requests: %v", <-pluginMock.errorChan)
+	}
+
+	// Ensuring that the test secret has been cached - written to etcd and read back by cacher.
+	pollErr := wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+		_, err := test.restClient.CoreV1().Secrets(test.ns.Name).Get(test.secret.Name, metav1.GetOptions{ResourceVersion: "0"})
+		return err == nil, nil
+	})
+
+	if pollErr == wait.ErrWaitTimeout {
+		pluginMock.cleanUp()
+		test.cleanUp()
+		t.Fatalf("failed to retrieve secret from cache within the alloted time period: %d", wait.ForeverTestTimeout)
+	}
+
+	return pluginMock, test
+}
+
+func mustReadSecret(t *testing.T, test *transformTest, bypassCache bool) {
+	// ResourceVersion 0 implies cache will be checked, "" implies that cache will be bypassed.
+	resourceVersion := "0"
+	if bypassCache {
+		resourceVersion = ""
+	}
+
+	s, err := test.restClient.CoreV1().Secrets(test.ns.Name).Get(test.secret.Name, metav1.GetOptions{ResourceVersion: resourceVersion})
+	if err != nil {
+		t.Fatalf("failed to get %s, error: %v", test.secret.Name, err)
+	}
 	if secretVal != string(s.Data[secretKey]) {
 		t.Fatalf("expected %s from KubeAPI, but got %s", secretVal, string(s.Data[secretKey]))
 	}
-	test.printMetrics()
 }
 
 func getDEKFromKMSPlugin(pluginMock *base64Plugin) ([]byte, error) {
-	// We expect KMS to already have seen an encryptRequest. Hence non-blocking call.
-	e, ok := <-pluginMock.encryptRequest
-
-	if !ok {
-		return nil, fmt.Errorf("failed to sense encryptRequest from KMS Plugin Mock")
+	select {
+	case e := <-pluginMock.encryptRequest:
+		return e.Plain, nil
+	case <-time.After(wait.ForeverTestTimeout):
```

Comment:
can this be ~5s instead?

Reply:
It probably could, but I  have seen some pretty significant time-outs in our tests.
Also, this was recommended by Tim.